libgsm: upstream Linux fixes

danielnachun opened the pull request:
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

danielnachun commented on Formula/libgsm.rb:
```diff
@@ -35,8 +41,12 @@ def install
-    system "make", "lib/libgsm.1.0.13.dylib",
-           "CC=#{ENV.cc}", "CCFLAGS=#{ENV.cflags}",
+    library = "libgsm.1.0.13.dylib"
+    on_linux { library = "libgsm.so" }
```
The fix in linuxbrew-core does not version the shared library on Linux like we do on macOS.  Should we do this on Linux too?

carlocab replied:
Probably, yes.

danielnachun replied:
I just noticed that this is already taken care of by the patch we are using (which is failing because of the Ruby error posted below).  However, the so version in the patch is out of date (1.0.18).  @iMichka I think we should move that patch from your repo to https://github.com/Homebrew/formula-patches so other maintainers can update it.

I also noticed that the dylib version for macOS is very outdated (1.0.13), and we should probably change the formula so that it fills in version automatically.  `libgsm` has no dependents so nothing else needs a revision bump if we do this.

carlocab replied:
We should maybe put an `@@SOVERSION@@` placeholder in the patch that we `inreplace` in the formula so that the patch doesn't need to be manually updated.

danielnachun commented:
Not sure what the license here should be.  It vaguely resembles `BSD`, but doesn't match the ones I've checked:
```
Copyright 1992, 1993, 1994 by Jutta Degener and Carsten Bormann,
Technische Universitaet Berlin

Any use of this software is permitted provided that this notice is not
removed and that neither the authors nor the Technische Universitaet Berlin
are deemed to have made any representations as to the suitability of this
software for any purpose nor are held responsible for any defects of
this software.  THERE IS ABSOLUTELY NO WARRANTY FOR THIS SOFTWARE.

As a matter of courtesy, the authors request to be informed about uses
this software has found, about bugs in this software, and about any
improvements that may be of general interest.

Berlin, 28.11.1994
Jutta Degener
Carsten Bormann

                                 oOo

Since the original terms of 15 years ago maybe do not make our
intentions completely clear given today's refined usage of the legal
terms, we append this additional permission:

      Permission to use, copy, modify, and distribute this software
      for any purpose with or without fee is hereby granted,
      provided that this notice is not removed and that neither
      the authors nor the Technische Universitaet Berlin are
      deemed to have made any representations as to the suitability
      of this software for any purpose nor are held responsible
      for any defects of this software.  THERE IS ABSOLUTELY NO
      WARRANTY FOR THIS SOFTWARE.

Berkeley/Bremen, 05.04.2009
Jutta Degener
Carsten Bormann
```

danielnachun commented:
Here is the debug output for the error I'm getting now.  I'm guessing there's some issue with using `on_linux`/`on_macos` blocks inside `patch do` block?
```
Error: An exception occurred within a child process:
  FrozenError: can't modify frozen String
/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libgsm.rb:47:in `block in install'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/extend/os/linux/on_os.rb:8:in `on_linux'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/libgsm.rb:47:in `install'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/debrew.rb:17:in `block in install'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/debrew.rb:90:in `debrew'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/debrew.rb:17:in `install'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/build.rb:178:in `block (2 levels) in install'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1281:in `block in brew'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:2415:in `block (2 levels) in stage'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/utils.rb:550:in `with_env'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:2414:in `block in stage'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/resource.rb:121:in `block (2 levels) in unpack'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/download_strategy.rb:115:in `chdir'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/download_strategy.rb:115:in `chdir'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/download_strategy.rb:102:in `stage'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/resource.rb:117:in `block in unpack'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/mktemp.rb:63:in `block in run'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/mktemp.rb:63:in `chdir'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/mktemp.rb:63:in `run'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/resource.rb:199:in `mktemp'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/resource.rb:116:in `unpack'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/resource.rb:91:in `stage'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/lib/ruby/2.6.0/forwardable.rb:230:in `stage'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:2394:in `stage'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1274:in `brew'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/build.rb:138:in `block in install'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/utils.rb:550:in `with_env'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/build.rb:133:in `install'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/build.rb:229:in `<main>'
```

carlocab commented on Formula/libgsm.rb:
```diff
@@ -35,8 +41,12 @@ def install
-           "CC=#{ENV.cc}", "CCFLAGS=#{ENV.cflags}",
+    library = "libgsm.1.0.13.dylib"
+    on_linux { library = "libgsm.so" }
+    ccflags = ENV.cflags
+    on_linux { ccflags << "-fPIC" }
+    system "make", "lib/#{library}",
+           "CC=#{ENV.cc}", "CCFLAGS=#{ccflags}",
```
```suggestion
    on_linux { ENV.append_to_cflags "-fPIC" }
    system "make", "lib/#{library}",
           "CC=#{ENV.cc}", "CCFLAGS=#{ENV.cflags}",
```

danielnachun commented:
I moved these patches over to formula-patches and changed them so that we can use the SOVERSION placeholder to put the right version in the Makefile at build time.  Moving the patches over also seemed to fix the syntax failure, so this is ready to go!

SMillerDev commented:
Can we also patch this upstream?

MikeMcQuaid commented:
> Can we also patch this upstream?

Yes please. A patch per-OS that's not submitted upstream is a pretty hard 👎🏻 from me, personally.

MikeMcQuaid commented on Formula/libgsm.rb:
```diff
@@ -35,7 +41,13 @@ def install
     # Dynamic library must be built first
-    system "make", "lib/libgsm.1.0.13.dylib",
+    inreplace "Makefile", "@@SOVERSION@@", version
+    library = "libgsm.#{version}.dylib"
+    on_linux do
+      library = "libgsm.so"
```
Why version it on one OS and not the other? Also, a helper that spits out `dylib` or `so` given an optional `version` would be good here, I think.

alebcay replied:
`shared_library` already accepts an optional `version` parameter. However it looks like the OS-specific code is needed here since the .dylib is versioned but the .so is not.

SMillerDev replied:
Why not submit a patch upstream to rectify that? That doesn't seem like something _anyone_ would be happy with.

MikeMcQuaid reviewed (commented):
Would like to see a little cleanup here.

MikeMcQuaid commented:
> Without the patch it only provides headers and a static library

This seems fine.

> so it would be hard to use as a dependency in another formula.

I disagree, it would just make it a `:build` dependency (which doesn't really matter for something that's unmaintained).

> carrying these patches would not be a major maintenance burden.

Can you 100% guarantee these patches introduce no regressions in functionality or security vulnerabilities? This is a somewhat rhetorical question; you cannot and this is (one of the multiple reasons) why we don't like carrying patches indefinitely in homebrew-core.

MikeMcQuaid commented:
@danielnachun have updated this based on https://github.com/Homebrew/brew/pull/11955

danielnachun commented:
@iMichka the consensus here at the moment is that we should not be carrying a patch to build a shared library for this.  If we go down that route, I will change this PR to delete the patches here, and also change that in linuxbrew-core.

carlocab commented:
This will break users of the shared library, including anything a user may have built on their own against Homebrew `libgsm`.

MikeMcQuaid commented:
> I say this is _more_ of a security risk than patching it.

I disagree. Our documented patching strategy also states that this is an approach we shouldn't be taken.

> We're not alone in this in this idea of wanting shared libraries. Many other package managers use have a shared library patch. In fact I genuinely tried and could not find one that doesn't patch it.

This again points towards the upstream software being unmaintained which means we should deprecate it rather than patch it indefinitely. I don't have strong opinions on whether deprecation or removing the patch is a better option but I'm strongly opposed to these patches.

> This sounds hard to enforce without audits so if that sounds like a maintenance nightmare, maybe the alternative is relinquishing it from homebrew-core entirely.

Sounds good.

> On the note about the OS conditionals: the patch could definitely be made OS-generic without much issue by moving OS conditionals inside the Makefile. No one has bothered before because package managers are usually not cross-platform and is thus a pointless exercise for them. Nix is the only one I can think of, and they just do short `sed` replacements to cover this so haven't bothered to make a full patch.

Another reason, to me, to point on these sort of patches not being things we should rely on.

danielnachun commented:
@Bo98 that would be great if you can take a look - it sounds like you've already got a pretty good idea of what to do!
